File: backend/github/pr_comment.py
from collections import defaultdict
import re

from backend.models import AIAnalysis, Finding

COMMENT_MARKER_PREFIX = "<!-- complypatch-ai:pr-comment:"


def build_comment_marker(head_sha: str) -> str:
    return f"{COMMENT_MARKER_PREFIX}{head_sha} -->"


def generate_pr_comment(
    score: int,
    findings: list[Finding],
    marker: str | None = None,
    ai_analysis: AIAnalysis | None = None,
) -> str:
    grouped: dict[str, list[Finding]] = defaultdict(list)
    for finding in findings:
        grouped[finding.severity].append(finding)

    comment = "## ComplyPatch AI Review\n\n"
    comment += f"**Compliance Score:** {score}/100\n\n"
    if ai_analysis and ai_analysis.status == "completed" and ai_analysis.riskScore is not None:
        comment += f"**AI Risk Score:** {ai_analysis.riskScore}/100"
        if ai_analysis.riskLevel:
            comment += f" ({ai_analysis.riskLevel})"
        comment += "\n\n"

    if not findings:
        comment += "No major compliance risks were detected in this scan.\n"
        return comment

    comment += "### Summary\n"
    comment += f"ComplyPatch AI detected {len(findings)} compliance/security finding(s).\n\n"

    for severity in ["Critical", "High", "Medium", "Low"]:
        severity_findings = grouped[severity]
        if not severity_findings:
            continue

        comment += f"### {severity} Findings\n\n"

        for index, finding in enumerate(severity_findings, start=1):
            comment += f"{index}. **{finding.title}** ({finding.ruleId})\n"
            comment += f"   - File: `{finding.file}:{finding.line}`\n"
            comment += f"   - Evidence: `{safe_inline(finding.evidence)}`\n"
            comment += f"   - Risk: {finding.impact}\n"
            comment += f"   - Suggested fix: {finding.fix}\n\n"

    comment += "### Recommendation\n"
    if ai_analysis and ai_analysis.status == "completed" and ai_analysis.suggestedRemediation:
        comment += f"{ai_analysis.suggestedRemediation}\n\n"
    comment += "Do not merge until Critical and High findings are reviewed and fixed.\n\n"
    comment += "_This is an automated compliance-readiness review, not legal certification._\n"
    if marker:
        comment += f"\n{marker}\n"

    return comment


def safe_inline(value: str) -> str:
    sanitized = redact_sensitive_value(value.replace("`", "'"))
    return sanitized[:180]


def redact_sensitive_value(value: str) -> str:
    sanitized = value
    sanitized = re.sub(r"sk-[A-Za-z0-9_-]{8,}", "sk-[redacted]", sanitized)
    sanitized = re.sub(
        r"(?i)(api[_-]?key|access[_-]?token|private[_-]?key|password)\s*=\s*['\"][^'\"]+['\"]",
        r"\1='[redacted]'",
        sanitized,
    )
    return sanitized

File: lib/github/pr-comment.ts
import type { Finding, FindingCounts } from "@/lib/scanner/types";

export function generatePrComment(
  score: number,
  findings: Finding[],
  options: { disclaimer?: string; findingCounts?: FindingCounts } = {}
) {
  const grouped = groupBySeverity(findings);
  const disclaimer = options.disclaimer ?? "This report provides compliance assistance, not legal certification.";

  let comment = `## ComplyPatch AI Review\n\n`;
  comment += `**Compliance Score:** ${score}/100\n\n`;
  comment += `_Demo note: live GitHub posting is mocked for this hackathon flow._\n\n`;

  if (findings.length === 0) {
    comment += `No major compliance risks were detected in this scan.\n`;
    comment += `\n_${disclaimer}_\n`;
    return comment;
  }

  comment += `### Summary\n`;
  comment += `ComplyPatch AI detected ${findings.length} compliance/security finding(s).\n\n`;
  if (options.findingCounts) {
    comment += `Critical: ${options.findingCounts["severity:Critical"] ?? 0}, `;
    comment += `High: ${options.findingCounts["severity:High"] ?? 0}, `;
    comment += `Medium: ${options.findingCounts["severity:Medium"] ?? 0}, `;
    comment += `Low: ${options.findingCounts["severity:Low"] ?? 0}\n\n`;
  }

  for (const severity of ["Critical", "High", "Medium", "Low"] as const) {
    const severityFindings = grouped[severity] ?? [];
    if (severityFindings.length === 0) continue;

    comment += `### ${severity} Findings\n\n`;

    severityFindings.forEach((finding, index) => {
      comment += `${index + 1}. **${finding.title}** (${finding.ruleId})\n`;
      comment += `   - File: \`${finding.file}:${finding.line}\`\n`;
      comment += `   - Category: ${finding.category}\n`;
      comment += `   - Evidence: \`${safeInline(finding.evidence)}\`\n`;
      comment += `   - Risk: ${finding.impact}\n`;
      comment += `   - Suggested fix: ${finding.fix}\n\n`;
    });
  }

  comment += `### Recommendation\n`;
  comment += `Do not merge until Critical and High findings are reviewed and fixed.\n\n`;
  comment += `_${disclaimer}_\n`;

  return comment;
}

function groupBySeverity(findings: Finding[]) {
  return findings.reduce<Record<string, Finding[]>>((acc, finding) => {
    acc[finding.severity] = acc[finding.severity] ?? [];
    acc[finding.severity].push(finding);
    return acc;
  }, {});
}

function safeInline(value: string) {
  return value.replace(/`/g, "'").slice(0, 180);
}
